File: client/src/app/services/player-service/player.service.ts
import { Injectable } from '@angular/core';
import { SocketService } from '@app/services/communication-socket/communication-socket.service';
import { BONUS, DEFAULT_ACTIONS, DEFAULT_ATTACK, DEFAULT_DEFENSE, DEFAULT_EVASIONS, DEFAULT_HP, DEFAULT_SPEED, ProfileType } from '@common/constants';
import { Avatar, Bonus, Player, Specs } from '@common/game';

@Injectable({
    providedIn: 'root',
})
export class PlayerService {
    player: Player;

    constructor(private socketService: SocketService) {
        this.resetPlayer();
        this.socketService = socketService;
    }

    createPlayer() {
        const playerSpecs: Specs = {
            life: this.player.specs.life,
            speed: this.player.specs.speed,
            attack: this.player.specs.attack,
            defense: this.player.specs.defense,
            attackBonus: this.player.specs.attackBonus,
            defenseBonus: this.player.specs.defenseBonus,
            movePoints: this.player.specs.speed,
            evasions: DEFAULT_EVASIONS,
            actions: DEFAULT_ACTIONS,
            nVictories: 0,
            nDefeats: 0,
            nCombats: 0,
            nEvasions: 0,
            nLifeTaken: 0,
            nLifeLost: 0,
            nItemsUsed: 0,
        };
        const player: Player = {
            name: this.player.name,
            socketId: this.socketService.socket.id || '',
            level: this.player.level,
            isActive: true,
            isEliminated: false,
            isObserver: false,
            avatar: this.player.avatar,
            specs: playerSpecs,
            inventory: [],
            position: { x: 0, y: 0 },
            initialPosition: { x: 0, y: 0 },
            turn: 0,
            visitedTiles: [],
            profile: ProfileType.NORMAL,
        };
        this.player = player;
    }

    setPlayer(player: Player): void {
        this.player = player;
    }

    setPlayerName(name: string): void {
        this.player.name = name.trim();
    }

    setPlayerAvatar(avatar: Avatar): void {
        this.player.avatar = avatar;
    }

    setPlayerLevel(level: number): void {
        this.player.level = level;
    }

    assignBonus(type: 'life' | 'speed'): void {
        if (type === 'life') {
            this.player.specs.life = DEFAULT_HP + BONUS;
            this.player.specs.speed = DEFAULT_SPEED;
        } else if (type === 'speed') {
            this.player.specs.speed = DEFAULT_SPEED + BONUS;
            this.player.specs.life = DEFAULT_HP;
        }
    }

    assignDice(type: 'attack' | 'defense'): void {
        if (type === 'attack') {
            this.player.specs.attackBonus = Bonus.D6;
            this.player.specs.defenseBonus = Bonus.D4;
        } else if (type === 'defense') {
            this.player.specs.attackBonus = Bonus.D4;
            this.player.specs.defenseBonus = Bonus.D6;
        }
    }

    resetPlayer(): void {
        const playerSpecs: Specs = {
            life: DEFAULT_HP,
            speed: DEFAULT_SPEED,
            attack: DEFAULT_ATTACK,
            defense: DEFAULT_DEFENSE,
            attackBonus: Bonus.D6,
            defenseBonus: Bonus.D4,
            evasions: DEFAULT_EVASIONS,
            movePoints: 0,
            actions: DEFAULT_ACTIONS,
            nVictories: 0,
            nDefeats: 0,
            nCombats: 0,
            nEvasions: 0,
            nLifeTaken: 0,
            nLifeLost: 0,
            nItemsUsed: 0,
        };
        const player: Player = {
            name: '',
            socketId: '',
            level: 1,
            isActive: true,
            isEliminated: false,
            isObserver: false,
            avatar: Avatar.Avatar1,
            specs: playerSpecs,
            inventory: [],
            position: { x: 0, y: 0 },
            initialPosition: { x: 0, y: 0 },
            turn: 0,
            visitedTiles: [],
            profile: ProfileType.NORMAL,
        };
        this.player = player;
    }
}

File: mobile/lib/services/player_service.dart
import 'package:flutter/foundation.dart';
import 'package:mobile/common/game.dart';
import 'package:mobile/common/map_types.dart';
import 'package:mobile/services/game_service.dart';
import 'package:mobile/utils/debug_logger.dart';

class PlayerService {
  factory PlayerService() => _instance;
  PlayerService._();
  static final PlayerService _instance = PlayerService._();

  final ValueNotifier<Player> notifier = ValueNotifier(
    Player(socketId: '', name: '', avatar: Avatar.avatar1, specs: Specs()),
  );

  Player get player => notifier.value;

  void setPlayer(Player newPlayer) {
    DebugLogger.log(
      'PlayerService: setting player ${newPlayer.name}',
      tag: 'PlayerService',
    );
    notifier.value = newPlayer;
  }

  void setPlayerFromJson(Map<String, dynamic> json) {
    try {
      final newPlayer = parsePlayer(json);
      setPlayer(newPlayer);
    } on Exception catch (e) {
      DebugLogger.log(
        'PlayerService: failed to parse player: $e',
        tag: 'PlayerService',
      );
    }
  }

  static Player parsePlayer(Map<String, dynamic> json) {
    final specsJson = json['specs'] as Map<String, dynamic>? ?? {};
    final inventoryJson = json['inventory'] as List<dynamic>? ?? [];
    final positionJson = json['position'] as Map<String, dynamic>?;
    final visitedJson = json['visitedTiles'] as List<dynamic>? ?? [];

    return Player(
      socketId: json['socketId'] as String? ?? '',
      name: json['name'] as String? ?? '',
      avatar: Avatar.values[(json['avatar'] as int? ?? 1) - 1],
      level: json['level'] as int? ?? 1,
      isActive: json['isActive'] as bool? ?? true,
      isEliminated: json['isEliminated'] as bool? ?? false,
      isGameWinner: json['isGameWinner'] as bool? ?? false,
      isObserver: json['isObserver'] as bool? ?? false,
      wasActivePlayer: json['wasActivePlayer'] as bool? ?? false,
      specs: Specs(
        life: specsJson['life'] as int? ?? 0,
        evasions: specsJson['evasions'] as int? ?? 0,
        speed: specsJson['speed'] as int? ?? 0,
        attack: specsJson['attack'] as int? ?? 0,
        defense: specsJson['defense'] as int? ?? 0,
        attackBonus:
            (specsJson['attackBonus'] as int? ?? 4) == 6 ? Bonus.d6 : Bonus.d4,
        defenseBonus:
            (specsJson['defenseBonus'] as int? ?? 6) == 6 ? Bonus.d6 : Bonus.d4,
        movePoints: specsJson['movePoints'] as int? ?? 0,
        actions: specsJson['actions'] as int? ?? 0,
        nVictories: specsJson['nVictories'] as int? ?? 0,
        nDefeats: specsJson['nDefeats'] as int? ?? 0,
        nCombats: specsJson['nCombats'] as int? ?? 0,
        nEvasions: specsJson['nEvasions'] as int? ?? 0,
        nLifeTaken: specsJson['nLifeTaken'] as int? ?? 0,
        nLifeLost: specsJson['nLifeLost'] as int? ?? 0,
        nItemsUsed: specsJson['nItemsUsed'] as int? ?? 0,
      ),
      inventory:
          inventoryJson
              .map((i) => GameService.parseItemCategory(i.toString()))
              .toList(),
      position:
          positionJson != null
              ? [Coordinate(positionJson['x'] as int, positionJson['y'] as int)]
              : [],
      turn: json['turn'] as int? ?? 0,
      visitedTiles:
          visitedJson
              .map((v) => Coordinate(v['x'] as int, v['y'] as int))
              .toList(),
      profile: GameService.parseProfileType(json['profile'] as String?),
    );
  }

  void clearPlayer() {
    DebugLogger.log('PlayerService: clearing player', tag: 'PlayerService');
    notifier.value = Player(
      socketId: '',
      name: '',
      avatar: Avatar.avatar1,
      specs: Specs(),
    );
  }
}
